pkg/controllers/updaterun/validation.go

```diff
@@ -0,0 +1,319 @@
+	// Remember the newly computed stage status.
+	newStageStatus := updateRun.Status.StagesStatus
+	// Make sure the number of stages in the clusterStagedUpdateRun are still the same.
+	if len(existingStageStatus) != len(newStageStatus) {
+		mismatchErr := fmt.Errorf("the number of stages in the clusterStagedUpdateRun has changed, new: %d, existing: %d", len(newStageStatus), len(existingStageStatus))
+		klog.ErrorS(mismatchErr, "The number of stages in the clusterStagedUpdateRun has changed", "clusterStagedUpdateRun", klog.KObj(updateRun))
+		return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, mismatchErr.Error())
+	}
+	// Make sure the stages in the updateRun are still the same.
+	for curStage := range existingStageStatus {
+		if existingStageStatus[curStage].StageName != newStageStatus[curStage].StageName {
+			mismatchErr := fmt.Errorf("index `%d` stage name in the clusterStagedUpdateRun has changed, new: %s, existing: %s", curStage, newStageStatus[curStage].StageName, existingStageStatus[curStage].StageName)
+			klog.ErrorS(mismatchErr, "The stage name in the clusterStagedUpdateRun has changed", "clusterStagedUpdateRun", klog.KObj(updateRun))
+			return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, mismatchErr.Error())
+		}
+		if len(existingStageStatus[curStage].Clusters) != len(newStageStatus[curStage].Clusters) {
+			mismatchErr := fmt.Errorf("the number of clusters in index `%d` stage has changed, new: %d, existing: %d", curStage, len(newStageStatus[curStage].Clusters), len(existingStageStatus[curStage].Clusters))
+			klog.ErrorS(mismatchErr, "The number of clusters in the stage has changed", "clusterStagedUpdateRun", klog.KObj(updateRun))
+			return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, mismatchErr.Error())
+		}
+		// Check that the clusters in the stage are still the same.
+		for j := range existingStageStatus[curStage].Clusters {
+			if existingStageStatus[curStage].Clusters[j].ClusterName != newStageStatus[curStage].Clusters[j].ClusterName {
+				mismatchErr := fmt.Errorf("the `%d` cluster in the `%d` stage has changed, new: %s, existing: %s", j, curStage, newStageStatus[curStage].Clusters[j].ClusterName, existingStageStatus[curStage].Clusters[j].ClusterName)
+				klog.ErrorS(mismatchErr, "The cluster in the stage has changed", "clusterStagedUpdateRun", klog.KObj(updateRun))
+				return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, mismatchErr.Error())
+			}
+		}
+
+		var err error
+		updatingStageIndex, lastFinishedStageIndex, err = determineUpdatingStage(curStage, updatingStageIndex, lastFinishedStageIndex, &existingStageStatus[curStage], updateRun)
+		if err != nil {
+			return -1, -1, err
+		}
+	}
+	return updatingStageIndex, lastFinishedStageIndex, nil
+}
+
+// determineUpdatingStage validates a stage status and looks for the updating stage and last finished stage.
+// It returns latest updatingStageIndex, lastFinishedStageIndex and any error encountered.
+func determineUpdatingStage(
```

Comment:
the function probably could use a bit more descriptive name

Reply:
Thanks, updated.